org/mozilla/jss/crypto/Algorithm.c

```diff
@@ -97,7 +102,10 @@ JSS_AlgInfo JSS_AlgTable[NUM_ALGS] = {
 /* 66 */    {CKM_AES_KEY_WRAP_PAD, PK11_MECH},
 /* 67 */    {CKM_SHA256_HMAC, PK11_MECH},
 /* 68 */    {CKM_SHA384_HMAC, PK11_MECH},
-/* 69 */    {CKM_SHA512_HMAC, PK11_MECH}
+/* 69 */    {CKM_SHA512_HMAC, PK11_MECH},
+
+/* CKM_AES_CMAC is new to NSS; some implementations might not yet have it. */
+/* 70 */    {CKM_AES_CMAC, PK11_MECH}
```

Comment:
Another question:
I assume by adding this mechanism , which nss now implements, is enough to make existing jni code in jss do this job for us? Just curious to take a look at what's going on down there. Is PK11MessageDigest.c where the magic is happening?

Reply:
Yep! That's some lovely magic.

So my understanding of PKCS#11 is a bit... rough, and Bob is definitely the one to ask these questions to. :)

But there's a few basic operations in PKCS#11: Sign, Verify, Encrypt, Decrypt, etc. Each of these correspond to a top-level function call in NSS: `PK11_Sign`, `PK11_Verify`, etc. So, once you've set up the context, you essentially get away with calling one of those methods, and then you'll have the result. But anyways... let's start at the top.

You create a new [`javax.crypto.Mac`](https://docs.oracle.com/javase/8/docs/api/javax/crypto/Mac.html) instance of type [`CmacAES`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/JSSProvider.java#L244-L257) (or one of its aliases). This gets backed by a [`JSSMacSpi`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/provider/javax/crypto/JSSMacSpi.java) instance of subclass [`CmacAES`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/provider/javax/crypto/JSSMacSpi.java#L121-L125). All this does is initialize the `JSSMacSpi` instance with the `CMACAlgorithm.AES` algorithm. This is an instance of [`org.mozilla.jss.crypto.CMACAlgorithm`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/crypto/CMACAlgorithm.java) with the correct PKCS#11 constant (er, proxies). [*]

Anyhow, so when you call `Mac.init`, it calls into [`PK11MessageDigest.initHMAC(...)`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.java#L35-L58), which creates the [PKCS#11 mechanism](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.c#L61), makes sure the key is enabled [for signing](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.c#L71), and then creates the [PKCS#11 context](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.c#L82). 

Each call to `Mac.update` then calls [`PK11_DigestOp`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.c#L127), and in the end [`PK11_DigestFinish`](https://github.com/cipherboy/jss/blob/add-cmac/org/mozilla/jss/pkcs11/PK11MessageDigest.c#L166) is called. And to finish; `PK11_DigestOp`/`PK11_DigestFinal` is merely a [nice wrapper](https://github.com/nss-dev/nss/blob/master/lib/pk11wrap/pk11cxt.c#L989-L1008) around all of those different PKCS#11 operations.

[*]: This is where I disagree with the construction of `org.mozilla.jss.crypto.Algorithm`: rather than handling PKCS#11-backed and SecOID-backed algorithms separately, they merge it into one big table of tuples (value, type). Then, they define pretty values like `Algorithm.CKM_AES_CMAC = 70`, which is an *index* and not the actual value. They could've instead made all PKCS#11-backed algorithms use, well, the underlying [`PKCS11Constants`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/pkcs11/PKCS11Constants.html) values and not made a redundant mapping (that class was added by me since its counterpart [in the JDK](http://www.docjar.com/docs/api/sun/security/pkcs11/wrapper/PKCS11Constants.html) was removed in JDK9, but that's well before our time)... But that's neither here nor there, and we have [`org.mozilla.jss.crypto.PKCS11Algorithm`](https://dogtagpki.github.io/jss/master/javadocs/org/mozilla/jss/crypto/PKCS11Algorithm.html) to map between them for us.

Reply:
Excellent info. I will approve since it looked good to me, just wanted a couple of clarifications. I just wanted to make sure I missed some other code that calls down into your new nss cmac code. Thanks!